tests/test_mcp_servers.py

```diff
@@ -1,100 +1,373 @@
-            provider_id="model-context-protocol",
-            mcp_endpoint={"uri":REMOTE_MCP_URL},
+    return logger
+
+
+def load_queries(file_path: str) -> List[Dict[str, str]]:
+    """Load query strings from a JSON file."""
+    try:
+        with open(file_path, 'r') as file:
+            data = json.load(file)
+
+        if not isinstance(data, dict) or 'queries' not in data:
+            raise ValueError(f"Invalid JSON format in {file_path}")
+
+        # Return full query objects with ID for better test identification
+        return data['queries']
+    except FileNotFoundError:
+        print(f"Query file not found: {file_path}")
+        return []
+    except json.JSONDecodeError:
+        print(f"Invalid JSON in file: {file_path}")
+        return []
+
+
+def get_query_id(query_obj):
+    """Extract an ID from a query object for better test identification."""
+    if isinstance(query_obj, dict) and 'id' in query_obj:
+        return query_obj['id']
+    elif isinstance(query_obj, dict) and 'query' in query_obj:
+        # Use first few words of query if no ID is available
+        words = query_obj['query'].split()[:5]
+        return '_'.join(words).lower().replace(',', '').replace('.', '')
+    return "unknown_query"
+
+
+def register_toolgroup_if_needed(
+    client: LlamaStackClient,
+    toolgroup_id: str,
+    mcp_url: str,
+    logger: logging.Logger
+) -> bool:
+    """Register a toolgroup if it doesn't exist. Returns success status."""
+    try:
+        registered_tools = client.tools.list()
+        registered_toolgroups = set(t.toolgroup_id for t in registered_tools)
+
+        logger.info(f"Available toolgroups: {registered_toolgroups}")
+
+        if toolgroup_id not in registered_toolgroups:
+            logger.info(f"Registering toolgroup: {toolgroup_id}")
+            client.toolgroups.register(
+                toolgroup_id=toolgroup_id,
+                provider_id="model-context-protocol",
+                mcp_endpoint={"uri": mcp_url},
             )
+            logger.info(f"Successfully registered toolgroup: {toolgroup_id}")
+        else:
+            logger.info(f"Toolgroup {toolgroup_id} is already registered")
 
-    logger.info(f"""Your Server has access the the following toolgroups:
-    {set(registered_toolgroups)}
-    """)
+        return True
+    except Exception as e:
+        logger.error(f"Failed to register toolgroup {toolgroup_id}: {e}")
+        return False
 
-    for prompt in user_prompts:
-        # Create simple agent with tools
-        agent = Agent(
-            client,
-            model=model,
-            instructions = """You are a helpful assistant. You have access to a number of tools.
+
+def execute_query(
+    client: LlamaStackClient,
+    prompt: str,
+    model: str,
+    toolgroup_id: str,
+    logger: logging.Logger,
+    instructions: Optional[str] = None,
+    max_tokens: int = 4096
+) -> Dict[str, Any]:
+    """Execute a single query."""
+    if instructions is None:
```

Comment:
so if a user query requires a specific system prompt to be passed for a given model, how can we handle that? does is it have to be provided in the `queries.json` or can it be passed in the `run_test()` function somehow?